### DISCUSSION TITLE
[WASM] Accessing the local file system?

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#9641] [Q&A] [answered]
I'm trying to access **my PC's file system**. Guess, my principle question is: **is that possible at all using UNO WASM?**

1) I'm using FolderPicker - which actually brings up the the Windows UI to select/create a folder - but I'm not sure if the folder/path returned would be on my local PC or somewhere in browser's IndexedDB? Please see below : I'm aware of [this ](https://platform.uno/docs/articles/features/windows-storage-pickers.html)but it's not clear to me whether this code 
```
(await new Windows.Storage.Pickers.FolderPicker
    {
       FileTypeFilter			= { "*" },
       SuggestedStartLocation	= PickerLocationId.Desktop,
    }
    .PickSingleFolderAsync())?.Path
```
... is supposed to return an absolute path on the local PC or not?
[FolderPicker.zip](https://github.com/unoplatform/uno/files/9438407/FolderPicker.zip)

2) Next I'm trying to create/access [StorageFiles ](https://platform.uno/docs/articles/features/windows-storage.html)and again I'm not clear - although I provided a path like "c:\tmp\muell.txt" - if that really goes to my PC's local file system or to the browser's IndexedDB?

This is latest Windows 11 + latest Chrome.

Could you please provide clarity if it's possible to access the local PC's file system at all using (UNO) WASM? Thanks in advance

**Comment by jeromelaban** [comment]
Please see [the documentation here](https://platform.uno/docs/articles/features/windows-storage-pickers.html?q=picke#webassembly).

**Reply by DierkDroth** [reply]
Thanks @jeromelaban 

Sorry, but I'm still not clear: in my case the FolderPicker clearly shows me the folder structure of my local PC.

I then would assume that `(await new Windows.Storage.Pickers.FolderPicker()....PickSingleFolderAsync())?.Path` returns an absolute path to a real folder on my PC, correct? 
However, that's not the case (see repo). Is this a bug?

(There are more follow up issues which I'm holding off until I clearly understand the case above).

Thanks

**Comment by DierkDroth** [comment]
@jeromelaban I'm sorry, but I'm banging my head since I don't understand how this works and what I'm seeing here. I really needed your help (and - yes - I have read the link you provided - in fact I even provided that link in my original post).

Before I threw in a couple of (may be) stupid questions let's start with a simple one. Where is the content of this file going (in WASM)?
```
File.WriteAllText(Path.Combine(ApplicationData.Current.LocalFolder.Path, "Muell.txt"), "this is a test");
```
Note: `ApplicationData.Current.LocalFolder.Provider.Id == "computer"` and `ApplicationData.Current.LocalFolder.Path == "/local"`. 

I would expect that I then would find a file '/local/muell.txt' in Application -> Storage -> IndexedDB -> '/local ...' -> FILE_DATA. 
**But there is no key '/local/muell.txt'.**

What am I missing?

**Reply by jeromelaban** [reply]
Did you read this part as well ? https://platform.uno/docs/articles/features/file-management.html#webassembly-file-system

**Reply by DierkDroth** [reply]
Absolutely. This is why I came up with my simple sample

**Reply by jeromelaban** [reply]
When writing to the file system with `File.WriteAllText` the content is written to the indexdb, but you need to wait for the indexdb to be initialized as described in the article above, otherwise, it may be overwritten. It is also written back to the indexdb regularly, not immediately.

**Reply by DierkDroth** [reply]
@jeromelaban thanks for clarification. I got that working.

Now to my real question (sorry if this sounds stupid, but I really don't know): is there any way I could write/read random files to/from my local hard drive using WASM or is this just not 'permitted' by the brower's security model and all I could do is writing data to the browser's IndexedDB storage (or a browser's memory filesystem, which is not of interest to me)?

**Reply by jeromelaban** [reply]
It is not permitted by the browser unless it's allowed by the user. The files in those folders can only be manipulated through the `Windows.Storage` APIs, and not by `File.IO` APIs.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban 

Please see modified repo below: [FolderPicker is supposed to work](https://platform.uno/docs/articles/features/windows-storage-pickers.html) with the file system of my local PC, no? However, it's not returning the **absolute** path.

Is that expected?

If so, then how else could I then get the correct, absolute path of the folder selected? 
Note: I needed to persist the path so I would not have the query the user every time. AFAIK this only could be done by the **correct, absolute** path.

[FolderPicker.zip](https://github.com/unoplatform/uno/files/9476165/FolderPicker.zip)

**Reply by DierkDroth** [reply]
@jeromelaban would you mind reviewing the repo I attached? 

Is the observed behavior a bug? Should I rather file an issue report for further investigation?

Thanks in advance

**Comment by MartinZikmund** [comment, marked as answer]
Unfortunately this is not possible. WASM file pickers use the https://wicg.github.io/file-system-access/ API, which does not offer any way to access the absolute path the picked file/folder for security reasons. If you require persisting access to the file/folder over multiple launches, we have to implement the `FutureAccessList` API that will use local storage to persist the native JS file handle to be reusable across launches. If that it something you need, please upvote https://github.com/unoplatform/uno/issues/5343
